src/main/java/duke/Parser.java

```diff
@@ -0,0 +1,205 @@
+package duke;
+import java.util.List;
+import java.util.Arrays;
+import java.time.LocalDate;
+import java.time.format.DateTimeParseException;
+
+/**
+ * A Parser is the logic handling agent of the user's commands and database inputs when loading from it.
+ */
+public class Parser {
+
+    /**
+     * The powerful thing of a parser is it does not rely on any external
+     * agent to perform the processing of inputs, just with pure logic.
+     */
+    public Parser() {
+    }
+
+    /**
+     * Processes every command line entered by the user. Users are able to enter
+     * bye, list, done x, find key, delete x, todo description, event description /at time,
+     * deadline description /by time, and any other inputs are deemed as invalid entries.
+     * @param command The string of command entered by the user.
+     * @param taskList The TaskList to process CRUD operations based on user input.
+     * @return An output denoting the success / failure of an operation.
+     * @throws DukeTerminateException Shuts down the Duke by entering "bye".
+     * @throws InsufficientInputException Not enough to arguments to find/delete/done.
+     * @throws InvalidIndexException Wrong indexes from TaskList's delete and done operations.
+     * @throws InvalidTodoException Invalid format of Task entered by the user.
+     * @throws InvalidEventException Invalid format of Event entered by the user.
+     * @throws InvalidDeadlineException Invalid format of Deadline entered by the user.
+     * @throws InvalidInputException Invalid commands entered by the user.
+     */
+    public String processCommand(String command, TaskList taskList) throws DukeTerminateException,
+            InsufficientInputException, InvalidIndexException, InvalidTodoException, InvalidEventException, InvalidDeadlineException,
+            InvalidInputException {
+        String output = "";
+        if (command.equals("bye")) { // exit
+                throw new DukeTerminateException();
```

Comment:
Think of another way to stop the program when the user enters 'bye'. It is technically a valid 'command', so you should not be throwing an exception!

<details>
  <summary>Hint</summary>
  Just check the input string in Duke class
</details>